src/choose_gossip_peer_strategy.rs

```diff
@@ -0,0 +1,306 @@
+use crdt::ReplicatedData;
+use rand::distributions::{IndependentSample, Weighted, WeightedChoice};
+use rand::thread_rng;
+use result::{Error, Result};
+use signature::PublicKey;
+use std;
+use std::collections::HashMap;
+
+pub const DEFAULT_WEIGHT: u32 = 1;
+
+pub trait ChooseGossipPeerStrategy {
+    fn choose_peer(&self, options: Vec<&ReplicatedData>) -> Result<ReplicatedData>;
+}
+
+pub struct ChooseRandomPeerStrategy<'a> {
+    random: &'a Fn() -> u64,
+}
+
+impl<'a> ChooseRandomPeerStrategy<'a> {
+    pub fn new(random: &'a Fn() -> u64) -> Self {
+        ChooseRandomPeerStrategy { random }
+    }
+}
+
+impl<'a> ChooseGossipPeerStrategy for ChooseRandomPeerStrategy<'a> {
+    fn choose_peer(&self, options: Vec<&ReplicatedData>) -> Result<ReplicatedData> {
+        if options.len() < 1 {
+            return Err(Error::CrdtTooSmall);
+        }
+
+        let n = ((self.random)() as usize) % options.len();
+        Ok(options[n].clone())
+    }
+}
+
+pub struct ChooseWeightedPeerStrategy<'a> {
+    remote: &'a HashMap<PublicKey, u64>,
+    external_liveness: &'a HashMap<PublicKey, HashMap<PublicKey, u64>>,
+    get_stake: &'a Fn(PublicKey) -> f64,
+}
+
+impl<'a> ChooseWeightedPeerStrategy<'a> {
+    pub fn new(
+        remote: &'a HashMap<PublicKey, u64>,
+        external_liveness: &'a HashMap<PublicKey, HashMap<PublicKey, u64>>,
+        get_stake: &'a Fn(PublicKey) -> f64,
```

Comment:
This seems unnecessarily generic. Why not make it a method of this struct?

Reply:
I left it in mostly for flexible testing purposes, I was afraid that whoever added a real implementation later would break my test cases and be forced to fix them. This way, I can define the function in my tests without worrying about adjusting for future changes.